File: Data Structures & Algorithms/reorder-linked-list/submission-4.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode *slow, *fast;
        slow = fast = head;
        
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        
        ListNode *prev, *curr, *nxt;
        prev = slow;
        nxt = curr = slow->next;
        prev->next = nullptr;

        while (curr) {
            nxt = nxt->next;
            curr->next = prev;
            prev = curr;
            curr = nxt;
        }
        
        ListNode *l, *r, *temp;
        l = head;
        r = prev;
        
        while (l != r) {
            temp = l->next;
            l->next = r;
            l = temp;
            temp = r->next;
            if (!temp) break;
            r->next = l;
            r = temp;
        }
    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        length = 0
        
        iterator = head
        while iterator:
            length += 1
            iterator = iterator.next
        if length == 1: 
            head.next = None
            return

        n_node = length - n
        dummy_node = ListNode(next = head)
        iterator = dummy_node
        for i in range(n_node): iterator = iterator.next
        remove = iterator.next
        iterator.next = remove.next

        return dummy_node.next